test/performance/odp_random.c

```diff
@@ -0,0 +1,376 @@
+	printf("\n"
+	       "random data performance test\n"
+	       "\n"
+	       "Usage: odp_random [options]\n"
+	       "\n"
+	       "  -t, --threads Number of worker threads (default %u)\n"
+	       "  -s, --size    Size of buffer in bytes (default %u)\n"
+	       "  -r, --rounds  Number of test rounds (default %u)\n"
+	       "                Rounded down to nearest multiple of 8\n"
+	       "  -h, --help    This help\n"
+	       "\n",
+	       options_def.num_threads, options_def.size, options_def.rounds);
+}
+
+static int parse_options(int argc, char *argv[])
+{
+	int opt;
+	int long_index;
+	int ret = 0;
+
+	static const struct option longopts[] = {
+		{ "threads", required_argument, NULL, 't' },
+		{ "size", required_argument, NULL, 's' },
+		{ "rounds", required_argument, NULL, 'r' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	static const char *shortopts = "+t:s:r:h";
+
+	options = options_def;
+
+	while (1) {
+		opt = getopt_long(argc, argv, shortopts, longopts, &long_index);
+
+		if (opt == -1)
+			break;
+
+		switch (opt) {
+		case 't':
+			options.num_threads = atol(optarg);
+			break;
+		case 's':
+			options.size = atol(optarg);
+			break;
+		case 'r':
+			options.rounds = atol(optarg);
+			break;
+		case 'h':
+			/* fall through */
+		default:
+			print_usage();
+			ret = -1;
+			break;
+		}
+	}
+
+	if (options.size < 1) {
+		ODPH_ERR("Invalid size: %" PRIu32 "\n", options.size);
+		return -1;
+	}
+
+	return ret;
+}
+
+const char *shm_name = "odp_random_test";
+
+typedef struct test_shm_t {
+	odp_barrier_t barrier;
+	odp_random_kind_t type;
+	uint64_t nsec[ODP_THREAD_COUNT_MAX];
+} test_shm_t;
+
+static test_shm_t *shm_lookup(void)
+{
+	test_shm_t *shm = NULL;
+	odp_shm_t shm_hdl = odp_shm_lookup(shm_name);
+
+	if (shm_hdl != ODP_SHM_INVALID)
+		shm = (test_shm_t *)odp_shm_addr(shm_hdl);
+
+	return shm;
+}
+
+static int test_random(void *p)
+{
+	(void)p;
+
+	uint8_t *buf, *data;
+	const unsigned long page = 4 * KB;
+	odp_time_t start;
+	uint64_t nsec;
+	test_shm_t *shm = shm_lookup();
+
+	if (!shm) {
+		ODPH_ERR("Failed to look up shm %s\n", shm_name);
+		exit(EXIT_FAILURE);
+	}
+
+	/* One extra page for alignment, another one for offset. */
+	buf = (uint8_t *)malloc(options.size + page * 2);
+
+	if (!buf) {
+		ODPH_ERR("Memory allocation failed.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	/* Align to start of page. */
+	data = (uint8_t *)(((uintptr_t)buf + (page - 1)) & ~(page - 1));
+
+	odp_barrier_wait(&shm->barrier);
+	start = odp_time_local();
+
+	for (uint32_t i = 0; i < options.rounds; i++) {
+		uint32_t pos = 0;
+
+		while (pos < options.size) {
+			int32_t n = odp_random_data(
+				data + pos, options.size - pos, shm->type);
+
+			if (n < 0) {
+				ODPH_ERR("odp_random_data() failed\n");
+				exit(EXIT_FAILURE);
+			}
+
+			pos += n;
+		}
+	}
+
+	nsec = odp_time_diff_ns(odp_time_local(), start);
+	shm->nsec[odp_thread_id()] = nsec;
+	free(buf);
+
+	return 0;
+}
+
+static int test_random_test(void *p)
```

Comment:
This is an almost exact copy of test_random(). It would be better to have single function that would take a parameter that tells if the normal or test random data function is to be tested.

Reply:
Should the conditional branch then be in the inner loop, the outer loop, or outside the loops? In the latter two cases, a third of the function would be repeated anyway.

Reply:
In the inner loop (a conditional or call through a function pointer or whatever the exact mechanism is). I would guess the overhead is negligible, but if you are worried about it, then you could force the common function to be always inlined.

Reply:
I'm not, but @psavol was worried about using a function pointer in some other performance test, probably crc. I'm also not worried about repeating a few lines of test code, so anything's fine with me.

Reply:
I tried the conditional, and in my environment in the worst case (1 byte length test data) the measured throughput decreases by 3.4%. If that's acceptable, I can push the change. I left the code duplication as is for now.